WordPress/Classes/Services/MediaService.m

```diff
@@ -356,24 +356,49 @@ - (void)updateMultipleMedia:(NSArray<Media *> *)mediaObjects
 
 }
 
-- (NSError *)translateMediaUploadError:(NSError *)error {
-    NSError *newError = error;
-    if (error.domain == WPXMLRPCFaultErrorDomain) {
-        NSString *errorMessage = [error localizedDescription];
-        NSInteger errorCode = error.code;
-        NSMutableDictionary *userInfo = [NSMutableDictionary dictionaryWithDictionary:error.userInfo];
-        switch (error.code) {
-            case 500:{
-                errorMessage = NSLocalizedString(@"Your site does not support this media file format.", @"Message to show to user when media upload failed because server doesn't support media type");
-            } break;
-            case 401:{
-                errorMessage = NSLocalizedString(@"Your site is out of space for media uploads.", @"Message to show to user when media upload failed because user doesn't have enough space on quota/disk");                
-            } break;
+- (NSError *)customMediaUploadError:(NSError *)error remote:(id <MediaServiceRemote>)remote {
+    NSString *customErrorMessage = nil;
+    if ([remote isKindOfClass:[MediaServiceRemoteXMLRPC class]]) {
+        // For self-hosted sites we should generally pass on the raw system/network error message.
+        // Which should help debug issues with a self-hosted site.
+        if ([error.domain isEqualToString:WPXMLRPCFaultErrorDomain]) {
+            switch (error.code) {
+                case 500:{
+                    customErrorMessage = NSLocalizedString(@"Your site does not support this media file format.", @"Message to show to user when media upload failed because server doesn't support media type");
+                } break;
+                case 401:{
+                    customErrorMessage = NSLocalizedString(@"Your site is out of storage space for media uploads.", @"Message to show to user when media upload failed because user doesn't have enough space on quota/disk");
+                } break;
+            }
         }
-        userInfo[NSLocalizedDescriptionKey] = errorMessage;
-        newError = [[NSError alloc] initWithDomain:WPXMLRPCFaultErrorDomain code:errorCode userInfo:userInfo];
+    } else if ([remote isKindOfClass:[MediaServiceRemoteREST class]]) {
+        // For WPCom/Jetpack sites and NSURLErrors we should use a more general error message to encourage a user to try again,
+        // rather than raw NSURLError messaging.
+        if ([error.domain isEqualToString:NSURLErrorDomain]) {
+            switch (error.code) {
+                case NSURLErrorUnknown:
+                    // Unknown error, encourage the user to try again
+                    // Note: if support requests show up with this error message we can rule out known NSURLError codes
+                    customErrorMessage = NSLocalizedString(@"Media upload failed, an unknown error occurred, please try again.", @"Error message shown when a media upload fails for unknown reason and the user should try again.");
```

Comment:
See my other comment about "Media upload failed" – perhaps this should just be "An unknown error occurred. Please try again."